Lib/test/test_typing.py

```diff
@@ -447,6 +447,83 @@ def test_bad_var_substitution(self):
                     list[T][arg]
 
 
+class TypeVarSubstitutionTests(BaseTestCase):
+
+    def test(self):
+        T = TypeVar('T')
+        T1 = TypeVar('T1')
+        T2 = TypeVar('T2')
+        Ts = TypeVarTuple('Ts')
+        class A(Generic[T]): pass
+        class B(Generic[T1, T2]): pass
```

Comment:
Since generics for builtins use a different code path, should we repeat the tests for A and B using list and dict, respectively? Using a quick hack it looks like I found two that behave differently:
```
======================================================================
FAIL: test (test.test_typing.TypeVarSubstitutionTests.test) (alias='list[T, *tuple[int, ...]]', args='[int]', expected=<class 'TypeError'>)----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\gvanrossum\pr-32341\Lib\test\test_typing.py", line 518, in test
    with self.assertRaises(expected):
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: TypeError not raised

======================================================================
FAIL: test (test.test_typing.TypeVarSubstitutionTests.test) (alias='dict[T1, T2, *tuple[int, ...]]', args='[int, str]', expected=<class 'TypeError'>)----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\gvanrossum\pr-32341\Lib\test\test_typing.py", line 518, in test
    with self.assertRaises(expected):
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: TypeError not raised

----------------------------------------------------------------------
```

Reply:
Following up on myself:

Manually:
```
>>> list[T, *tuple[int, ...]]
list[~T, *tuple[int, ...]]
>>> list[T, *tuple[int, ...]][int]
list[int, *tuple[int, ...]]

>>> dict[T1, T2, *tuple[int, ...]]
dict[~T1, ~T2, *tuple[int, ...]]
>>> dict[T1, T2, *tuple[int, ...]][int, str]
dict[int, str, *tuple[int, ...]]
>>>
```

Honestly, the fewer `TypeError`s we get from such constructs the better -- our general philosophy is leaning towards only flagging the most egregious errors. For example, `list[T][int, int]` is rejected, but `list[T][42]` is not.

Reply:
Good point. I've added tests for `list`, `dict` and `tuple`, annotating where I think the current implementation is wrong bearing in mind that general philosophy.